### PR TITLE
SONAR-9858 Purge users and groups when deleting user, quality profile and organization

### PR DESCRIPTION
**Pull request by julienlancelot**

### CODE REVIEW

File: server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/QProfileWsSupport.java [hunk 1]
```diff
@@ -94,12 +100,52 @@ public QProfileDto getProfile(DbSession dbSession, QProfileReference ref) {
...
+    UserDto user = dbClient.userDao().selectActiveUserByLogin(dbSession, login);
+    checkFound(user, "User with login '%s' is not found'", login);
+    checkMembership(dbSession, organization, user);
+    return user;
+  }
+
+  public GroupDto getGroup(DbSession dbSession, OrganizationDto organization, String groupName) {
+    Optional<GroupDto> group = dbClient.groupDao().selectByName(dbSession, organization.getUuid(), groupName);
+    checkFoundWithOptional(group, "No group with name '%s' in organization '%s'", groupName, organization.getKey());
+    return group.get();
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/QProfileWsSupport.java]: MAJOR Call "group.isPresent()" before accessing the value. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3655)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* ![MAJOR][MAJOR] 1 major

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfileFactoryImpl.java [hunk 2]
```diff
@@ -127,6 +127,10 @@ public void delete(DbSession dbSession, Collection<QProfileDto> profiles) {
     db.defaultQProfileDao().deleteByQProfileUuids(dbSession, uuids);
     db.qualityProfileDao().deleteOrgQProfilesByUuids(dbSession, uuids);
 
+    // Permissions are only available on custom profiles
```
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfileFactoryImpl.java]: outside the scope of this PR: do you know why a factory class has a public delete method?

File: server/sonar-server/src/main/java/org/sonar/server/organization/ws/RemoveMemberAction.java [hunk 3]
```diff
@@ -98,6 +98,7 @@ private void removeMember(DbSession dbSession, OrganizationDto organization, Use
     String organizationUuid = organization.getUuid();
     dbClient.userPermissionDao().deleteOrganizationMemberPermissions(dbSession, organizationUuid, userId);
     dbClient.permissionTemplateDao().deleteUserPermissionsByOrganization(dbSession, organizationUuid, userId);
+    dbClient.qProfileEditUsersDao().deleteByUser(dbSession, user);
```
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/organization/ws/RemoveMemberAction.java]: shouldn't it be restricted to an organization? When you remove a member, you want to delete qprofile user permissions only for the given organization. Either use a `deleteByOrganizationAndUser` or get the organization profile to delete

File: server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfileFactoryImpl.java (hunk 2 above)
Comment (julienlancelot) [reply server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfileFactoryImpl.java]: It should be refactored later indeed

**Pull request merged**